### PR TITLE
RS-216: Time Series Blob Data: ReductStore vs. MongoDB

### PR DESCRIPTION
**Pull request by AnthonyCvn**

### CODE REVIEW

File: blog/2024-04-08-reductstore-vs-mongodb/index.md [hunk 1]
```diff
@@ -0,0 +1,254 @@
...
+        count = 0
+        for _ in range(BLOB_COUNT):
+            blob_id = fs.put(CHUNK, filename=f"blob_{datetime.now().timestamp()}")
+            data.insert_one({"time": datetime.now(), "blob_id": blob_id})
+            count += BLOB_SIZE
+
+    return count
+```
+
+### Write to Reduct
```
Comment (atimin) [inline blog/2024-04-08-reductstore-vs-mongodb/index.md:138]: ReductStore

File: blog/2024-04-08-reductstore-vs-mongodb/index.md [hunk 2]
```diff
@@ -0,0 +1,254 @@
...
+        async for rec in bucket.query("data", t1, t2, ttl=90):
+            count += len(await rec.read_all())
+
+        return count
+```
+
+The `read_all` method in ReductStore retrieves the entire binary content of a record. The `read` method, on the other hand, allows for reading chunks of data in a streaming fashion, providing flexibility in handling large binary payloads.
+
+## Performance Analysis: Insertion and Retrieval Speeds for Binary Data in MongoDB and ReductStore
+To evaluate the performance of MongoDB and ReductStore, we conducted a test using varying blob sizes ranging from 10 KB to 10 MB. The test measured the speed of writing and reading binary data on a Mac M1 with APFS SSD storage.
```
Comment (atimin) [inline blog/2024-04-08-reductstore-vs-mongodb/index.md:212]: NVMe

File: blog/2024-04-08-reductstore-vs-mongodb/index.md [hunk 3]
```diff
@@ -0,0 +1,254 @@
...
+
+For 10 KB chunks, MongoDB writes at a rate of 258 blobs per second, while ReductStore records a slightly lower rate at 223 blobs per second—a 14% decrease. However, when reading the same chunk size, ReductStore shows a marginal speed advantage with an increase of 4%.
+
+As chunk sizes escalate to 100 KB and beyond, up to 10 MB, ReductStore demonstrates significantly better performance over MongoDB. Write speeds improve from +36% for the former to +75% for the latter size tier. Similarly, read operations see improvements ranging from +34% up to +100%, highlighting ReductStore's efficiency in handling larger binary payloads.
+
+| Chunk Size | Operation | MongoDB, blob/s | ReductStore, blob/s | ReductStore, %    |
```
Comment (atimin) [inline blog/2024-04-08-reductstore-vs-mongodb/index.md:218]: Latest result with NVMe

File: blog/2024-04-08-reductstore-vs-mongodb/index.md [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+
+- **Replication features**: Replication in ReductStore is append-only, with label-based filtering options. This feature can contribute to an effective data reduction strategy by selectively replicating relevant blobs. MongoDB, on the other hand, offers replication sets for high availability and redundancy, ensuring data integrity and fault tolerance in distributed environments.
+
+- **Cost considerations**: ReductStore's pricing model is based on storage which aligns well with retention policies based on data volume for a predictable cost structure. MongoDB's pricing is based on instance size and usage, which can be more complex to estimate.
+
+## Conclusion
+In conclusion, ReductStore and MongoDB each offer distinct methods for managing and retrieving unstructured time series blob data. MongoDB requires explicit setup with GridFS, while ReductStore simplifies binary data storage into buckets. Performance analysis reveals that while MongoDB is competitive with smaller blob sizes, ReductStore provides superior throughput for larger payloads, making it a compelling choice for heavy-duty time series data scenarios.
```
Comment (atimin) [inline blog/2024-04-08-reductstore-vs-mongodb/index.md:242]: must be updated with the latest results

**Review by atimin — CHANGES_REQUESTED** [review]
It is perfect but needs being updated for NVMe.

File: blog/2024-04-08-reductstore-vs-mongodb/index.md [hunk 5]
```diff
@@ -0,0 +1,254 @@
...
+
+When selecting a database for blob storage, other criteria than performance must be taken into account:
+
+- **Retention policy**: both ReductStore and MongoDB offer distinct retention strategies. ReductStore uses disk utilization-based policies, essential for edge computing and IoT. MongoDB's time-series collections, on the other hand, can automatically purge documents after a set period of time, useful for managing data lifecycle when dealing with non-blob data such as numerical values or logs in time series scenarios.
+
+- **Replication features**: Replication in ReductStore is append-only, with label-based filtering options. This feature can contribute to an effective data reduction strategy by selectively replicating relevant blobs. MongoDB, on the other hand, offers replication sets for high availability and redundancy, ensuring data integrity and fault tolerance in distributed environments.
+
+- **Cost considerations**: ReductStore's pricing model is based on storage which aligns well with retention policies based on data volume for a predictable cost structure. MongoDB's pricing is based on instance size and usage, which can be more complex to estimate.
```
Comment (atimin) [inline blog/2024-04-08-reductstore-vs-mongodb/index.md:239]: Just checked sortly, their license is very similar to AGPL. You can use it for free as long as you don't provide MongoDB as service. If you do it, you should open your source code.

**Review by atimin — CHANGES_REQUESTED** [review]

File: blog/2024-04-08-reductstore-vs-mongodb/index.md (hunk 5 above)
Comment (AnthonyCvn) [reply blog/2024-04-08-reductstore-vs-mongodb/index.md:241]: @atimin I added a point "License Consideration" to be checked

File: blog/2024-04-08-reductstore-vs-mongodb/index.md (hunk 4 above)
Comment (AnthonyCvn) [reply blog/2024-04-08-reductstore-vs-mongodb/index.md]: I changed the conclusion accordingly. Let me know if it's ok

**Review by atimin — APPROVED** [review]
Perfect! Let's merge.

**Pull request merged**